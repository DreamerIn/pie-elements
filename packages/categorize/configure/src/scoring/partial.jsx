import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';
import Panel from './panel';
import PartialForCategory from './partial-for-category';
import Grid, { Cell } from './grid';

export class Partial extends React.Component {
  static propTypes = {
    classes: PropTypes.object.isRequired,
    className: PropTypes.string,
    partial: PropTypes.object.isRequired,
    onChange: PropTypes.func.isRequired,
    categories: PropTypes.array,
    correctResponse: PropTypes.array
  };

  static defaultProps = {};

  toggleEnabled = () => {
    const { partial, onChange } = this.props;
    partial.enabled = !partial.enabled;
    onChange(partial);
  };

  changeInnerPartial = p => {
    const { partial, onChange } = this.props;
    const index = partial.rules.find(r => r.category === p.category);
    if (index !== -1) {
      partial.rules.splice(index, 1, p);
      onChange(partial);
    }
  };
  render() {
    const {
      classes,
      className,
      partial,
      categories,
      correctResponse
    } = this.props;
    return (
      <div className={classNames(classes.partial, className)}>
        <Panel
          title="Partial Scoring"
          enabled={partial.enabled}
          onToggleEnabled={this.toggleEnabled}
        >
          <Grid headings={['Category', 'Scoring']} size={'40% 1fr'}>
            {categories.map((c, index) => (
              <React.Fragment key={index}>
                <Cell html={c.label} />
                <Cell>
                  <PartialForCategory
                    key={index}
                    category={c}
                    correctResponse={correctResponse.find(
                      r => r.category === c.id
                    )}
                    onChange={this.changeInnerPartial}
                    partial={partial.rules.find(r => r.category === c.id)}
                  />
                </Cell>
              </React.Fragment>
            ))}
          </Grid>
        </Panel>
      </div>
    );
  }
}
const styles = theme => ({
  partial: {}
});
export default withStyles(styles)(Partial);
