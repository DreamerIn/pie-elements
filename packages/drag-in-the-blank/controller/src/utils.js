import reduce from 'lodash/reduce';
import find from 'lodash/find';

export const getAllCorrectResponses = ({ correctResponse, alternateResponses }) => {
  return reduce(correctResponse || {}, (obj, val, key) => {
    obj[key] = [val];

    if (alternateResponses && alternateResponses[key]) {
      obj[key] = [
        ...obj[key],
        ...alternateResponses[key]
      ];
    }

    return obj;
  }, {});
};
