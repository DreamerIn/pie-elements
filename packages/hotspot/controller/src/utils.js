import isEqual from 'lodash/isEqual';
import isEmpty from 'lodash/isEmpty';

export const getCorrectResponse = (choices) => choices
  .filter(c => c.correct)
  .map(c => ({ id: c.id }))
  .sort();

export const isResponseCorrect = (question, session) => {
  const { shapes: { rectangles, polygons } } = question;
  const choices = [...rectangles, ...polygons];
  let correctResponse = getCorrectResponse(choices);

  if (!session || isEmpty(session)) {
    return false;
  }

  if (session.answers && session.answers.length) {
    return isEqual((session.answers || []).sort(), correctResponse);
  } else if (!(correctResponse && correctResponse.length)) {
    return true;
  }
  return false;
};
