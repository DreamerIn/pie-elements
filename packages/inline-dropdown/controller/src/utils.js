import isEqual from 'lodash/isEqual';
import reduce from 'lodash/reduce';
import find from 'lodash/find';

export const isResponseCorrect = (question, session) => {
  if (!question || !session || !session.value) {
    return false;
  }

  let correctResponse = reduce(question.choices, (acc, area, key) => {
    const correctChoice = find(area, c => c.correct);

    acc[key] = correctChoice.value;

    return acc;
  }, {});

  return isEqual((session.value || {}), correctResponse);
};
