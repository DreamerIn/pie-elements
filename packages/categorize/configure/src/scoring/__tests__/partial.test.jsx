import { shallow } from 'enzyme';
import React from 'react';
import { Partial } from '../partial';
import { callsOnChange as _callsOnChange } from '../../__tests__/utils';

describe('Partial', () => {
  let w;
  let onChange = jest.fn();
  const wrapper = extras => {
    const defaults = {
      classes: {},
      className: 'className',
      onChange,
      partial: {
        enabled: true,
        rules: [{ count: 1, percent: 20 }]
      },
      categories: [],
      correctResponse: []
    };
    const props = { ...defaults, ...extras };
    return shallow(<Partial {...props} />);
  };

  describe('snapshot', () => {
    it('renders', () => {
      w = wrapper();
      expect(w).toMatchSnapshot();
    });
  });
  describe('logic', () => {
    const callsOnChange = function() {
      const args = Array.prototype.slice.call(arguments);
      const first = args.shift();
      const startArgs =
        typeof first === 'string' ? [wrapper(), first] : [first];
      _callsOnChange.apply(null, startArgs.concat(args));
    };
    describe('toggleEnabled', () => {
      callsOnChange(
        wrapper({ partial: { enabled: false } }),
        'toggleEnabled',
        () => {
          expect(onChange).toBeCalledWith(
            expect.objectContaining({ enabled: true })
          );
        }
      );

      callsOnChange('toggleEnabled', () => {
        expect(onChange).toBeCalledWith(
          expect.objectContaining({ enabled: false })
        );
      });
    });

    describe('changeInnerPartial', () => {
      callsOnChange('changeInnerPartial', { count: 1, percent: 10 }, () => {
        expect(onChange).toBeCalledWith(
          expect.objectContaining({ rules: [{ count: 1, percent: 10 }] })
        );
      });
    });
  });
});
