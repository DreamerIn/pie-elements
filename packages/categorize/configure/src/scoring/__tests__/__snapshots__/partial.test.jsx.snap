// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Partial snapshot renders 1`] = `
<div
  className="className"
>
  <WithStyles(Panel)
    enabled={true}
    onToggleEnabled={[Function]}
    title="Partial Scoring"
  >
    <WithStyles(Grid)
      headings={
        Array [
          "Category",
          "Scoring",
        ]
      }
      size="40% 1fr"
    />
  </WithStyles(Panel)>
</div>
`;
